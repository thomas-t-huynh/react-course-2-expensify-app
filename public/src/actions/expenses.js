import uuid from 'uuid';

export const addExpenses = (
     {
          description ='',
          note = '', 
          amount = 0,
          createdAt = 0
         } = {}
    ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

export const removeExpense = ( { idToRemove } = {} ) => ({
    type: 'REMOVE_EXPENSE',
    idToRemove
});

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});