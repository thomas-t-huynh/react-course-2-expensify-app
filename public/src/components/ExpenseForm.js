import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { connect } from 'react-redux';

export class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        };
    }
    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        e.persist();
        this.setState(() => ({ note: e.target.value }));
    };


    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }; 
    };

    onFocusChange = ( { focused } ) => {
        this.setState(() => ({ calenderFocused: focused }))
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.note) {
            this.setState(() => ({ error: 'Please provide a title and a bloggle.' }))
        } else if (!this.props.isAuthenticated) {
            this.setState(() => ({ error: 'Please sign in to bloggle.' }))
        } else {
            this.setState(() => ({ 
                description: '',
                note: '',
                createdAt: '',
                error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
            console.log('submitted');
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <div className="text-button-flex">
                        <input 
                            type="text" 
                            placeholder="Title"
                            autoFocus
                            className="text-title"
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                        <button className="button" >Bloggle</button>
                    </div>
                    <textarea
                        placeholder="Start bloggle-ing"
                        value={this.state.note}
                        className="textarea"
                        onChange={this.onNoteChange}
                    >
                    </textarea>
            </form>
        )
    }
}

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
})


export default connect(mapStateToProps , undefined)(ExpenseForm);