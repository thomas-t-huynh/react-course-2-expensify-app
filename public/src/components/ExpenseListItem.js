import React from 'react';
import { Link } from 'react-router-dom';


const ExpenseListItem = ( { id , description , amount , createdAt } ) => (

    <div>
        <div> 
            <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
            <p>{amount} - {createdAt}</p>
            
        </div>
    </div>
);


export default ExpenseListItem;

