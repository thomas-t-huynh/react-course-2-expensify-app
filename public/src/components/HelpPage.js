import React from 'react';


const HelpPage = () => (
    <div>help Page</div>
);

export default HelpPage;