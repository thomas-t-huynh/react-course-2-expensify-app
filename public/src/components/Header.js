import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout , startLoginGoogle , startLoginFacebook } from '../actions/auth';


export const Header = ( {startLogout, startLoginGoogle , startLoginFacebook , isAuthenticated }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/">
                    <h1>Bloggle</h1>
                </Link>
                {isAuthenticated  ? <button className="button button--link" onClick={startLogout}>Logout</button> : <button className="button button--link" onClick={startLoginGoogle}>Log in</button>}
            </div>
        </div>
    </header>
);

const mapStateToProps = (state) => {
    return {
        isAuthenticated: !!state.auth.uid
    }
}

const mapDispatchToProps = (dispatch) => ({
    startLogout:  () => dispatch(startLogout()),
    startLoginGoogle: () => dispatch(startLoginGoogle()),
    startLoginFacebook: () => dispatch(startLoginFacebook())
});



export default connect(mapStateToProps, mapDispatchToProps)(Header);

