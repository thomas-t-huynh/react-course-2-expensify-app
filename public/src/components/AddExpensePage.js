import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpenses } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    
    onSubmit = ( expense ) => {
        this.props.startAddExpenses(expense);
        this.props.history.push('/dashboard');
    };

    render() {
       return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add expense</h1>
                    </div>
                </div>
                    <div className="content-container">
                        <ExpenseForm
                            onSubmit={this.onSubmit}
                        />
                    </div>    
            </div>
       );
    };
};

const mapDispatchToProps = (dispatch) => ({ startAddExpenses: (expense) => dispatch(startAddExpenses(expense)) });
        

export default connect(undefined , mapDispatchToProps)(AddExpensePage);
