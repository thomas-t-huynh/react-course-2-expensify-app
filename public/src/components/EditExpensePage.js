import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense , startRemoveExpense } from '../actions/expenses';
import RemoveModal from './RemoveModal';

export class EditExpensePage extends React.Component {

    state = {
        expenseInRemoval: false
    };

    setExpenseInRemoval = () => {
        this.setState(() => ( { expenseInRemoval: true } ))
    };

    clearExpenseInRemoval = () => {
        this.setState(() => ( { expenseInRemoval: false } ))
    };

   onSubmit = ( expense ) => {
        this.props.startEditExpense(this.props.expense.id , expense);
        this.props.history.push('/dashboard');
         
   };

   onRemove = () => {
       this.props.startRemoveExpense( {idToRemove: this.props.expense.id} );
       this.props.history.push('/dashboard');
       this.setState(() => ({ expenseInRemoval: false }))
   };

   
   render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense} 
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.setExpenseInRemoval}>Remove Expense</button>
                    <RemoveModal
                         expenseName={this.props.expense.description}
                         expenseInRemoval={this.state.expenseInRemoval}
                         onRemove={this.onRemove}
                         clearExpenseInRemoval={this.clearExpenseInRemoval}
                    />
                </div>
                    
            </div>
        );
    };
};

const mapStateToProps = (state , props) => {
    return {
        expense: state.expenses.find((expense) =>  expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch , props) => ({ 
    startEditExpense: (id, updates) => dispatch(startEditExpense(id , updates)),
    startRemoveExpense: (idToRemove) => dispatch(startRemoveExpense(idToRemove))
});



export default connect(mapStateToProps , mapDispatchToProps)(EditExpensePage);