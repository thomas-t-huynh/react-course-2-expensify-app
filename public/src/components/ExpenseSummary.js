import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/Expenses-total';
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import { startAddExpenses } from '../actions/expenses';

export class ExpenseSummary extends React.Component {  

    onSubmit = (expense) => {
        this.props.startAddExpenses(expense);
    }


    render() {
        return (
                <div className="page-header">
                    <div className="content-container">
                        <ExpenseForm onSubmit={this.onSubmit} />
                    </div>
                </div>
            );
    };
};



const mapDispatchToProps = (dispatch) => ({ startAddExpenses: (expense) => dispatch(startAddExpenses(expense)) });

export default connect(undefined , mapDispatchToProps)(ExpenseSummary);