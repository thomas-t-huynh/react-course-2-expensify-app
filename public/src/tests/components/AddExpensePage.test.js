import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let addExpenses, history, wrapper;

beforeEach(() => {
    addExpenses = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage addExpenses={addExpenses} history={history} />);
});

test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpenses).toHaveBeenLastCalledWith(expenses[2]);
});