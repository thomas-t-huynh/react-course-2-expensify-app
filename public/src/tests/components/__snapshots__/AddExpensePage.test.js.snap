// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    Create expense Page
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
